### PR TITLE
Add collaborative learning scripts

### PR DESCRIPTION
**Pull request by DamithaSenevirathne**

### CODE REVIEW

File: src_python/org_agg.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+        self.training_cycles = 0
+
+        self.stop_flag = False
+
+        # List of sockets for select.select()
+        self.sockets_list = []
+        self.clients = {}
+        self.client_ids = {}
+
+        # Craete Aggregator socket
```
Comment (miyurud) [inline src_python/org_agg.py:62]: Change `Crate` to `Create`?

File: src_python/org_agg.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+    def update_model(self,new_weights,num_examples):
+        self.partition_sizes.append(num_examples)
+        self.weights.append(num_examples * new_weights)
+
+        logging.info(str(self.partition_sizes))
+        logging.info(str(self.weights))
+
+        if len(self.weights) == self.NUM_CLIENTS:
+
+            avg_weight = sum(self.weights) / sum(self.partition_sizes)
```
Comment (miyurud) [inline src_python/org_agg.py:79]: Shouldn't we check `sum(self.partition_sizes) > 0` here?

File: src_python/org_agg.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            
+            logging.info("___________________________________________________ Training round %s done ______________________________________________________", self.training_cycles)
+        
+    def send_model(self, client_socket):
+
+        if self.ROUNDS == self.training_cycles:
+            self.stop_flag = True
+
+        weights = np.array(self.model_weights)
+        weights_path = "./weights/agg" + str(self.training_cycles) + ".npy"
```
Comment (miyurud) [inline src_python/org_agg.py:99]: Where do we store these weights? on JasmineGraph server's source folder?

File: src_python/org_agg.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+                    
+                        weights_path = "./weights/agg_a" + str(self.training_cycles) + ".npy"
+                        np.save(weights_path,weights)
+
+                        num_examples = message["NUM_EXAMPLES"]   #// commented
+                        self.client_ids[notified_socket] = client_id #// commented
+                    
+                    logging.info('Recieved model from client-%s at %s:%s',client_id, *self.clients[notified_socket])
+                    logging.info('Recieved model from client at %s:%s', *self.clients[notified_socket])
```
Comment (miyurud) [inline src_python/org_agg.py:178]: This second message seems part of the first message. I do not see any point of having this second log here.

File: src_python/org_agg.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+        'partition_id',
+        'num_orgs',
+        'num_rounds',
+        'IP',
+        'PORT'
+        ]
+
+    args = dict(zip(arg_names, sys.argv[1:]))
+
+    logging.warning('####################################### New Training Session #######################################')
```
Comment (miyurud) [inline src_python/org_agg.py:202]: Why is this message a warning? can't it be just info message?

File: src_python/org_agg.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    logging.info('aggregator started , graph ID %s, number of clients %s, number of rounds %s',args['graph_id'],args['num_orgs'],args['num_rounds'])
+
+    if 'IP' not in args.keys()  or args['IP'] == 'localhost':
+        args['IP'] = "localhost"
+
+    if 'PORT' not in args.keys():
+        args['PORT'] = 5000
+
+    path_nodes = args['path_nodes'] + args['graph_id'] + '_nodes_' + args['partition_id'] + ".csv"
+    nodes = pd.read_csv(path_nodes,index_col=0)
```
Comment (miyurud) [inline src_python/org_agg.py:212]: From where do we read this file? We should have read this from JasmineGraph data directory. Not from the JasmineGraph server software's folder.

File: src_python/org_server.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    server = Server(model_weights=model_weights,org_id=args['org_id'],rounds=int(args['num_rounds']),num_clients=int(args['num_clients']),ip=args['IP'],port=int(args['port']))
+    
+    logging.info('Federated training started!')
+    start = timer()
+    server.run()
+    end = timer()
+
+    elapsed_time = end -start
+    logging.info('Federated training done!')
+    logging.info('Training report : Elapsed time %s seconds, graph ID %s, number of clients %s, number of rounds %s',elapsed_time,args['org_id'],args['num_clients'],args['num_rounds'])
```
Comment (miyurud) [inline src_python/org_server.py:269]: Add a newline at the end of the file

File: src_python/org_agg.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    
+    logging.info('Federated training started!')
+
+    start = timer()
+    aggregator.run()
+    end = timer()
+
+    elapsed_time = end -start
+    logging.info('Federated training done!')
+    logging.info('Training report : Elapsed time %s seconds, graph ID %s, number of clients %s, number of rounds %s',elapsed_time,args['graph_id'],args['num_orgs'],args['num_rounds'])
```
Comment (miyurud) [inline src_python/org_agg.py:238]: Add a newline at the end of the file

**Review by miyurud — CHANGES_REQUESTED** [review]
Please address the comments I put in this PR

File: src_python/org_agg.py (hunk 1 above)
Comment (DamithaSenevirathne) [reply src_python/org_agg.py]: Fixed in 1875888487e22c8e17cbadd8acc558b28d98ebf0

File: src_python/org_agg.py (hunk 2 above)
Comment (DamithaSenevirathne) [reply src_python/org_agg.py:80]: Fixed in 1875888487e22c8e17cbadd8acc558b28d98ebf0

File: src_python/org_agg.py (hunk 3 above)
Comment (DamithaSenevirathne) [reply src_python/org_agg.py]: Added into separate weights folder inside jaminegraph. Changes in 1875888487e22c8e17cbadd8acc558b28d98ebf0

File: src_python/org_agg.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+                    if message is False:
+                        self.sockets_list.remove(notified_socket)
+                        del self.clients[notified_socket]
+                        continue
+                    else:
+                        client_id = message['ORG_ID']
+                        weights = message['WEIGHTS']
+
+                    
+                        weights_path = "./weights/agg_a" + str(self.training_cycles) + ".npy"
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply src_python/org_agg.py]: Updated in 1875888487e22c8e17cbadd8acc558b28d98ebf0

File: src_python/org_agg.py (hunk 4 above)
Comment (DamithaSenevirathne) [reply src_python/org_agg.py]: Removed in 1875888487e22c8e17cbadd8acc558b28d98ebf0

File: src_python/org_agg.py (hunk 5 above)
Comment (DamithaSenevirathne) [reply src_python/org_agg.py]: Updated in 1875888487e22c8e17cbadd8acc558b28d98ebf0

File: src_python/org_agg.py (hunk 6 above)
Comment (DamithaSenevirathne) [reply src_python/org_agg.py:213]: `    std::string command = path + "python3 org_server.py " + graphID+ " " + utils.getJasmineGraphProperty("org.jasminegraph.fl_clients")
                                    + " " + utils.getJasmineGraphProperty("org.jasminegraph.fl.epochs") `
we input these values as parameter from jasminegraph worker

Comment (DamithaSenevirathne) [reply src_python/org_agg.py:213]: These parameters (paths) can be only updated/changed by the jasminegraph conf file

File: src_python/org_server.py (hunk 7 above)
Comment (DamithaSenevirathne) [reply src_python/org_server.py]: FIxed in 1875888487e22c8e17cbadd8acc558b28d98ebf0

File: src_python/org_agg.py (hunk 3 above)
Comment (DamithaSenevirathne) [reply src_python/org_agg.py]: Changed file location to localstore 7018408aafed215ed25a1fa8237e25b9f4c978cf

File: src_python/org_agg.py (hunk 2 above)
Comment (miyurud) [reply src_python/org_agg.py:80]: Done

File: src_python/org_agg.py (hunk 1 above)
Comment (miyurud) [reply src_python/org_agg.py]: Done

File: src_python/org_agg.py (hunk 2 above)
Comment (miyurud) [reply src_python/org_agg.py:80]: Done

File: src_python/org_agg.py (hunk 3 above)
Comment (miyurud) [reply src_python/org_agg.py]: Done

File: src_python/org_agg.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (miyurud) [reply src_python/org_agg.py]: Done

File: src_python/org_agg.py (hunk 4 above)
Comment (miyurud) [reply src_python/org_agg.py]: Done

File: src_python/org_agg.py (hunk 5 above)
Comment (miyurud) [reply src_python/org_agg.py]: Done

File: src_python/org_agg.py (hunk 6 above)
Comment (miyurud) [reply src_python/org_agg.py:213]: Done

File: src_python/org_server.py (hunk 7 above)
Comment (miyurud) [reply src_python/org_server.py]: Done

File: src_python/org_agg.py (hunk 8 above)
Comment (miyurud) [reply src_python/org_agg.py]: Done

**Pull request merged**